File: mosaic-ts/src/cli/commands/janus.ts
/**
 * Phase 6 CLI: ``pnpm dev janus``.
 *
 * JANUS meta-weighting over the 7 PRISM regime cohorts:
 *   - run:     full daily cycle (weights + regime + blended recs), persisted
 *   - weights: cohort weights + 30d accuracy
 *   - regime:  regime signal (dominant cohort + concentration)
 *   - history: recent janus_runs (weight-drift over time)
 */

import type { Command } from "commander";
import pc from "picocolors";
import { BridgeApi, BridgeClient, RpcError } from "../../bridge/index.js";
import { pad } from "../_format.js";

interface DateOpts {
  date?: string;
  window?: string;
}

interface HistoryOpts {
  days?: string;
}

export function registerJanus(program: Command): void {
  const cmd = program
    .command("janus")
    .description("JANUS meta-weighting over the 7 PRISM regime cohorts (Phase 6).");

  cmd
    .command("run")
    .description("Full daily cycle: cohort weights + regime + blended recommendations.")
    .option("--date <date>", "As-of date (YYYY-MM-DD, default today)")
    .option("--window <n>", "Rolling accuracy window in days (default 30)")
    .action(async (opts: DateOpts) => {
      await withApi(async (api) => {
        const r = await api.janusRunDaily(dateParams(opts));
        console.log(pc.bold(`\njanus run -- ${r.date}`));
        printRegime(r.regime);
        printWeights(r.cohort_weights, r.cohort_accuracy);
        console.log(
          pc.cyan(
            `\n  blended: ${r.blended_recommendations.length} ` +
              `(${r.contested_tickers.length} contested)`,
          ),
        );
        for (const b of r.blended_recommendations.slice(0, 10)) {
          const flag = b.contested ? pc.yellow(" [contested]") : "";
          console.log(
            `  ${pad(b.direction, 6)} ${pad(b.ticker, 12)} ${b.blended_weight_pct.toFixed(1)}%${flag}`,
          );
        }
      });
    });

  cmd
    .command("weights")
    .description("Cohort weights + 30-day accuracy (no blend).")
    .option("--date <date>", "As-of date (YYYY-MM-DD, default today)")
    .option("--window <n>", "Rolling accuracy window in days (default 30)")
    .action(async (opts: DateOpts) => {
      await withApi(async (api) => {
        const r = await api.janusGetWeights(dateParams(opts));
        console.log(pc.bold(`\njanus weights -- ${r.date}`));
        printWeights(r.cohort_weights, r.cohort_accuracy);
      });
    });

  cmd
    .command("regime")
    .description("Regime signal (dominant cohort + concentration).")
    .option("--date <date>", "As-of date (YYYY-MM-DD, default today)")
    .option("--window <n>", "Rolling accuracy window in days (default 30)")
    .action(async (opts: DateOpts) => {
      await withApi(async (api) => {
        const r = await api.janusRegime(dateParams(opts));
        console.log(pc.bold(`\njanus regime -- ${r.date ?? ""}`));
        printRegime(r);
      });
    });

  cmd
    .command("history")
    .description("Recent JANUS runs (weight drift over time).")
    .option("--days <n>", "Number of runs to show (default 30)")
    .action(async (opts: HistoryOpts) => {
      await withApi(async (api) => {
        const days = opts.days ? Number.parseInt(opts.days, 10) : 30;
        const { history } = await api.janusGetHistory({ days });
        console.log(pc.bold(`\njanus history -- last ${days}`));
        if (history.length === 0) {
          console.log(pc.dim("  no runs recorded"));
          return;
        }
        console.log(
          pc.cyan(
            `\n  ${pad("date", 12)} ${pad("dominant", 18)} ${pad("conc", 8)} blended/contested`,
          ),
        );
        console.log(pc.dim(`  ${"─".repeat(56)}`));
        for (const h of history) {
          console.log(
            `  ${pad(h.date, 12)} ${pad(h.dominant_cohort ?? "-", 18)} ` +
              `${pad((h.concentration ?? 0).toFixed(3), 8)} ${h.n_blended}/${h.n_contested}`,
          );
        }
      });
    });
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function dateParams(opts: DateOpts): { date?: string; window_days?: number } {
  return {
    ...(opts.date ? { date: opts.date } : {}),
    ...(opts.window ? { window_days: Number.parseInt(opts.window, 10) } : {}),
  };
}

function printRegime(r: {
  dominant_cohort: string | null;
  regime_label: string;
  concentration: number;
  concentration_state?: string;
}): void {
  console.log(
    `  regime: ${pc.cyan(r.regime_label)} ` +
      `(dominant=${r.dominant_cohort ?? "-"}, ` +
      `concentration=${r.concentration.toFixed(3)} ${r.concentration_state ?? ""})`,
  );
}

function printWeights(
  weights: Record<string, number>,
  accuracy: Record<string, { hit_rate: number; sharpe: number; n: number }>,
): void {
  console.log(
    pc.cyan(`\n  ${pad("cohort", 18)} ${pad("weight", 8)} ${pad("hit", 7)} ${pad("sharpe", 8)} n`),
  );
  console.log(pc.dim(`  ${"─".repeat(50)}`));
  const sorted = Object.entries(weights).sort((a, b) => b[1] - a[1]);
  for (const [cohort, w] of sorted) {
    const a = accuracy[cohort] ?? { hit_rate: 0, sharpe: 0, n: 0 };
    console.log(
      `  ${pad(cohort, 18)} ${pad(`${(w * 100).toFixed(1)}%`, 8)} ` +
        `${pad(`${(a.hit_rate * 100).toFixed(0)}%`, 7)} ${pad(a.sharpe.toFixed(2), 8)} ${a.n}`,
    );
  }
}

async function withApi(fn: (api: BridgeApi) => Promise<void>): Promise<void> {
  const client = new BridgeClient();
  const api = new BridgeApi(client);
  try {
    await client.start();
    await fn(api);
  } catch (err) {
    if (err instanceof RpcError) {
      console.error(pc.red(`bridge error [${err.code}]: ${err.message}`));
    } else {
      console.error(pc.red(`error: ${(err as Error).message}`));
    }
    const tail = client.stderrTail.trim();
    if (tail) {
      console.error(pc.dim("\n--- bridge stderr (tail) ---"));
      console.error(pc.dim(tail.slice(-2000)));
    }
    process.exitCode = 1;
  } finally {
    await client.close();
  }
}

File: mosaic/bridge/handlers/janus.py
"""``janus.*`` JSON-RPC handlers (Plan §11.7 / Phase 6).

Exposes the JANUS meta-weighting layer to the TS front-end:

    * janus.run_daily   -- full cycle: weights + regime + blended recs, persisted
    * janus.get_weights -- cohort weights + 30d accuracy only (no blend)
    * janus.regime      -- regime signal only
    * janus.get_history -- recent janus_runs rows (TUI weight-drift chart)
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from ..protocol import INVALID_PARAMS, RpcError
from ..registry import method

_DEFAULT_WINDOW = 30


def _store():
    from mosaic.scorecard import get_store

    return get_store()


def _cohorts_and_configs():
    from mosaic.prism.cohorts import COHORT_CONFIGS

    return list(COHORT_CONFIGS.keys()), COHORT_CONFIGS


def _today() -> str:
    return datetime.now(timezone.utc).date().isoformat()


def _opt_date(params: dict) -> str:
    d = params.get("date")
    if d is not None and not isinstance(d, str):
        raise RpcError(INVALID_PARAMS, "'date' must be a string (YYYY-MM-DD)")
    return d or _today()


def _opt_window(params: dict) -> int:
    w = params.get("window_days", _DEFAULT_WINDOW)
    if not isinstance(w, int) or isinstance(w, bool) or w < 1:
        raise RpcError(INVALID_PARAMS, "'window_days' must be a positive integer")
    return w


@method("janus.run_daily")
def janus_run_daily(params: dict[str, Any]) -> dict[str, Any]:
    """Full daily JANUS cycle (weights + regime + blend), persisted."""
    from mosaic.janus import run_daily

    date = _opt_date(params)
    window = _opt_window(params)
    cohorts, configs = _cohorts_and_configs()
    now_iso = f"{date}T00:00:00+00:00"
    return run_daily(_store(), cohorts, date, now_iso=now_iso,
                     window_days=window, cohort_configs=configs)


@method("janus.get_weights")
def janus_get_weights(params: dict[str, Any]) -> dict[str, Any]:
    """Cohort weights + accuracy only (no recommendation blend)."""
    from mosaic.janus import compute_cohort_weights

    date = _opt_date(params)
    window = _opt_window(params)
    cohorts, _ = _cohorts_and_configs()
    weights, accuracy = compute_cohort_weights(
        _store(), cohorts, f"{date}T00:00:00+00:00", window
    )
    return {
        "date": date,
        "cohort_weights": {k: round(v, 4) for k, v in weights.items()},
        "cohort_accuracy": {
            k: {"hit_rate": round(v["hit_rate"], 4), "sharpe": round(v["sharpe"], 4), "n": v["n"]}
            for k, v in accuracy.items()
        },
    }


@method("janus.regime")
def janus_regime(params: dict[str, Any]) -> dict[str, Any]:
    """Regime signal only (dominant cohort + concentration)."""
    from mosaic.janus import compute_cohort_weights, regime_signal

    date = _opt_date(params)
    window = _opt_window(params)
    cohorts, configs = _cohorts_and_configs()
    weights, _ = compute_cohort_weights(_store(), cohorts, f"{date}T00:00:00+00:00", window)
    return {"date": date, **regime_signal(weights, configs)}


@method("janus.get_history")
def janus_get_history(params: dict[str, Any]) -> dict[str, Any]:
    """Recent janus_runs rows (newest first)."""
    days = params.get("days", _DEFAULT_WINDOW)
    if not isinstance(days, int) or isinstance(days, bool) or days < 1:
        raise RpcError(INVALID_PARAMS, "'days' must be a positive integer")
    return {"history": _store().get_janus_history(days)}
